src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -11,207 +12,171 @@ namespace System.Buffers
             }
+
+            return new SequencePosition(startSegment, startIndex + count);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal SequencePosition Seek(SequencePosition start, SequencePosition end, long count, bool checkEndReachable = true)
         {
-            int startIndex = start.Index;
-            int endIndex = end.Index;
-            SequenceType type = GetSequenceType();
-
-            startIndex = GetIndex(startIndex);
-            endIndex = GetIndex(endIndex);
+            GetTypeAndIndices(start.Index, end.Index, out SequenceType type, out int startIndex, out int endIndex);
 
-            switch (type)
+            object startSegment = start.Segment;
+            bool notInRange = endIndex - startIndex < count;
+            if (type == SequenceType.MemoryList)
             {
-                case SequenceType.MemoryList:
-                    if (start.Segment == end.Segment && endIndex - startIndex >= count)
-                    {
-                        // end.Index >= count + Index and end.Index is int
-                        return new SequencePosition(start.Segment, startIndex + (int)count);
-                    }
-                    return SeekMultiSegment((IMemoryList<byte>)start.Segment, startIndex, (IMemoryList<byte>)end.Segment, endIndex, count, checkEndReachable);
-
-                case SequenceType.OwnedMemory:
-                case SequenceType.Array:
-                    if (endIndex - startIndex >= count)
-                    {
-                        // end.Index >= count + Index and end.Index is int
-                        return new SequencePosition(start.Segment, startIndex + (int)count);
-                    }
-
-                    ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
-                    return default;
-
-                default:
-                    ThrowHelper.ThrowInvalidOperationException_UnexpectedSegmentType();
-                    return default;
+                object endSegment = end.Segment;
+                if (notInRange || startSegment != endSegment)
+                {
+                    return SeekMultiSegment(startSegment, startIndex, endSegment, endIndex, count, checkEndReachable);
+                }
             }
+            else if (notInRange)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
+            }
+
+            return new SequencePosition(startSegment, startIndex + (int)count);
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private static SequencePosition SeekMultiSegment(IMemoryList<byte> start, int startIndex, IMemoryList<byte> end, int endPosition, long count, bool checkEndReachable)
+        private static SequencePosition SeekMultiSegment(object startSegment, int startIndex, object endSegment, int endPosition, long count, bool checkEndReachable)
         {
+            var current = (IMemoryList<T>)startSegment;
```

Comment:
There is a random mix between `var` vs explicit variable types. Which should be used?

Reply:
explicit I think in corefx

Reply:
If the type can be inferred from the code line var can be used.